### PR TITLE
toolbox/emitter: add a prototype of spec-emitter

### PR DESCRIPTION
**Pull request by safl**
See the README.rst for usage on using this code-emitter prototype.

Here is what the is currently emitted with a limited model and emitter-templates, but it catches the gist of it:

```c
/**
 * Command Set Identifiers
 *
 * @enum xnvme_spec_csi
 */
enum xnvme_spec_csi {
	XNVME_SPEC_CSI_NVM = 0x0, ///< NVM Command Set Identifier
	XNVME_SPEC_CSI_KVS = 0x1, ///< KVS Command Set Identifier
	XNVME_SPEC_CSI_ZNS = 0x2, ///< ZNS Command Set Identifier
};/**
 * Log Page Identifiers
 *
 * @enum xnvme_spec_lpi
 */
enum xnvme_spec_lpi {
	XNVME_SPEC_LPI_ERROR                       = 0x1,  ///< tbd
	XNVME_SPEC_LPI_HEALTH_INFORMATION          = 0x2,  ///< tbd
	XNVME_SPEC_LPI_FIRMWARE_SLOT               = 0x3,  ///< tbd
	XNVME_SPEC_LPI_CHANGED_NS_LIST             = 0x4,  ///< tbd
	XNVME_SPEC_LPI_COMMAND_EFFECTS_LOG         = 0x5,  ///< tbd
	XNVME_SPEC_LPI_DEVICE_SELF_TEST            = 0x6,  ///< tbd
	XNVME_SPEC_LPI_TELEMETRY_HOST_INITIATED    = 0x7,  ///< tbd
	XNVME_SPEC_LPI_TELEMETRY_CTRLR_INITIATED   = 0x8,  ///< tbd
	XNVME_SPEC_LPI_ASSYMETRIC_NAMESPACE_ACCESS = 0xc,  ///< tbd
	XNVME_SPEC_LPI_LOG_DISCOVERY               = 0x70, ///< tbd
	XNVME_SPEC_LPI_RESERVATION_NOTIFICATION    = 0x80, ///< tbd
	XNVME_SPEC_LPI_SANITIZE_STATUS             = 0x81, ///< tbd
	XNVME_SPEC_LPI_VENDOR_SPECIFIC_START       = 0xc0, ///< tbd
	XNVME_SPEC_LPI_VENDOR_SPECIFIC_END         = 0xff, ///< tbd
};
/**
 * ADM command set opcodes
 *
 * @enum xnvme_spec_adm_opc
 */
enum xnvme_spec_adm_opc {
	XNVME_SPEC_ADM_OPC_DELETE_IO_SQ            = 0x0,  ///< Delete I/O Submission Queue
	XNVME_SPEC_ADM_OPC_CREATE_IO_SQ            = 0x1,  ///< Create I/O Submission Queue
	XNVME_SPEC_ADM_OPC_GET_LOG_PAGE            = 0x2,  ///< Get Log Page
	XNVME_SPEC_ADM_OPC_DELETE_IO_CQ            = 0x4,  ///< Delete I/O Completion Queue
	XNVME_SPEC_ADM_OPC_CREATE_IO_CQ            = 0x5,  ///< Create I/O Completion Queue
	XNVME_SPEC_ADM_OPC_IDENTIFY                = 0x6,  ///< Identify
	XNVME_SPEC_ADM_OPC_ABORT                   = 0x8,  ///< Abort
	XNVME_SPEC_ADM_OPC_SFEAT                   = 0x9,  ///< Set Features
	XNVME_SPEC_ADM_OPC_GFEAT                   = 0xa,  ///< Get Features
	XNVME_SPEC_ADM_OPC_ASYNC_EVENT_REQUEST     = 0xc,  ///< Async Event Request
	XNVME_SPEC_ADM_OPC_NS_MANAGEMENT           = 0xd,  ///< Namespace Management
	XNVME_SPEC_ADM_OPC_FIRMWARE_COMMIT         = 0x10, ///< Firmware Commit
	XNVME_SPEC_ADM_OPC_FIRMWARE_IMAGE_DOWNLOAD = 0x11, ///< Firmware Image Download
	XNVME_SPEC_ADM_OPC_DEVICE_SELF_TEST        = 0x14, ///< Device Self-test
	XNVME_SPEC_ADM_OPC_NS_ATTACHMENT           = 0x15, ///< Namespace Attachment
	XNVME_SPEC_ADM_OPC_KEEP_ALIVE              = 0x16, ///< Keep Alive
	XNVME_SPEC_ADM_OPC_DSEND                   = 0x19, ///< Directive Send
	XNVME_SPEC_ADM_OPC_DRECV                   = 0x1a, ///< Directive Receive
	XNVME_SPEC_ADM_OPC_VIRTUALIZATION_MGMT     = 0x1c, ///< Virtualization Management
	XNVME_SPEC_ADM_OPC_MI_SEND                 = 0x1d, ///< NVMe-MI Send
	XNVME_SPEC_ADM_OPC_MI_RECV                 = 0x1e, ///< NVMe-MI Receive
	XNVME_SPEC_ADM_OPC_CAPACITY_MGMT           = 0x20, ///< Capacity Management
	XNVME_SPEC_ADM_OPC_LOCKDOWN                = 0x24, ///< Lockdown
	XNVME_SPEC_ADM_OPC_DOORBELL_BUFFER_CONFIG  = 0x7c, ///< Doorbell buffer Config
	XNVME_SPEC_ADM_OPC_FABRICS_COMMANDS        = 0x7f, ///< Fabrics Commands
	XNVME_SPEC_ADM_OPC_FORMAT                  = 0x80, ///< Format
	XNVME_SPEC_ADM_OPC_SECURITY_SEND           = 0x81, ///< Security Send
	XNVME_SPEC_ADM_OPC_SECURITY_RECV           = 0x82, ///< Security Recv
	XNVME_SPEC_ADM_OPC_SANITIZE                = 0x84, ///< Sanitize
	XNVME_SPEC_ADM_OPC_GET_LBA_STATUS          = 0x86, ///< Get LBA Status
};

/**
 * NVM command set opcodes
 *
 * @enum xnvme_spec_nvm_opc
 */
enum xnvme_spec_nvm_opc {
	XNVME_SPEC_NVM_OPC_FLUSH                = 0x0,  ///< Flush
	XNVME_SPEC_NVM_OPC_WRITE                = 0x1,  ///< Write
	XNVME_SPEC_NVM_OPC_READ                 = 0x2,  ///< Read
	XNVME_SPEC_NVM_OPC_WRITE_UNCORRECTABLE  = 0x4,  ///< Write Uncorrectable
	XNVME_SPEC_NVM_OPC_COMPARE              = 0x5,  ///< Compare
	XNVME_SPEC_NVM_OPC_WRITE_ZEROES         = 0x8,  ///< Write Zeroes
	XNVME_SPEC_NVM_OPC_DATASET_MANAGEMENT   = 0x9,  ///< Dataset Management
	XNVME_SPEC_NVM_OPC_VERIFY               = 0xc,  ///< Verify
	XNVME_SPEC_NVM_OPC_RESERVATION_REGISTER = 0xd,  ///< Reservation Register
	XNVME_SPEC_NVM_OPC_RESERVATION_REPORT   = 0xe,  ///< Reservation Report
	XNVME_SPEC_NVM_OPC_RESERVATION_ACQUIRE  = 0x11, ///< Reservation Acquire
	XNVME_SPEC_NVM_OPC_RESERVATION_RELEASE  = 0x15, ///< Reservation Release
	XNVME_SPEC_NVM_OPC_SCOPY                = 0x19, ///< Simple Copy
};

/**
 * ZNS command set opcodes
 *
 * @enum xnvme_spec_zns_opc
 */
enum xnvme_spec_zns_opc {
	XNVME_SPEC_ZNS_OPC_MGMT_SEND = 0x79, ///< Zone Management Send
	XNVME_SPEC_ZNS_OPC_MGMT_RECV = 0x7a, ///< Zone Management Send
	XNVME_SPEC_ZNS_OPC_APPEND    = 0x7d, ///< Zone Append
};

/**
 * KVS command set opcodes
 *
 * @enum xnvme_spec_kvs_opc
 */
enum xnvme_spec_kvs_opc {
	XNVME_SPEC_KVS_OPC_STORE    = 0x1,  ///< Store
	XNVME_SPEC_KVS_OPC_RETRIEVE = 0x2,  ///< Retrieve
	XNVME_SPEC_KVS_OPC_DELETE   = 0x10, ///< Delete
	XNVME_SPEC_KVS_OPC_EXIST    = 0x14, ///< Exist
	XNVME_SPEC_KVS_OPC_LIST     = 0x6,  ///< List
};
/**
 * Status codes
 */
enum xnvme_spec_sc_generic {
	XNVME_SPEC_SC_SUCCESS             = 0x0,
	XNVME_SPEC_SC_INVALID_OPCODE      = 0x1,
	XNVME_SPEC_SC_INVALID_FIELD       = 0x2,
	XNVME_SPEC_SC_COMMAND_ID_CONFLICT = 0x3,
	XNVME_SPEC_SC_DATA_TRANSFER_ERROR = 0x4,
	XNVME_SPEC_SC_ABORTED_POWER_LOSS  = 0x5,
};

enum xnvme_spec_sc_command_specific {
	XNVME_SPEC_SC_COMPLETION_QUEUE_INVALID = 0x0,
};

enum xnvme_spec_sc_media_error {
	XNVME_SPEC_SC_INTERNAL_PATH_ERROR = 0x0,
};
/**
 * Status code types
 */
enum xnvme_spec_sc_type {
	XNVME_SPEC_SCT_GENERIC          = 0x0,
	XNVME_SPEC_SCT_COMMAND_SPECIFIC = 0x1,
	XNVME_SPEC_SCT_MEDIA_ERROR      = 0x2,
	XNVME_SPEC_SCT_PATH             = 0x3,
	XNVME_SPEC_SCT_VENDOR_SPECIFIC  = 0x7,
};
```

This approach seems doable for the following spec. entities:

* status codes
* status code types
* command opcodes
* log-page identifiers
* identify-command modifiers
* command-set identifiers
* feature identifiers
* sgl descriptor types
* sgl descriptor subtypes
* simple-copy formats
* zone-states
* zone-recv actions
* zone-recv-action specific fields
* basically... everything that are tabular listing of values in the NVMe specifications.

The code-emission for log-page-structs, command-structures, identify results etc. might also work for the general movement of the data-model to a code-emitter-template, but additional processing in that area is probably needed, or just some very verbose models.

The intent is that the model, that is, the ``.yaml`` files in the ``model`` directory should not be written by hand, but instead produced by https://github.com/linux-nvme/nvme-lint

### CODE REVIEW

**Comment by safl** [comment]
This is slowly growing into something, spinning it off for incubation at https://github.com/safl/yace
It will most likely turn into a tool that xNVMe, and other NVMe-related projects can make use of to generate headers, pretty-printers for structs, model-checking etc. so closing it of for now here.
